Automattic/pocket-casts-android
# End of Year: Story snapshot and share

| 📘 Project: #410 |
|:---:|

iOS PR: https://github.com/Automattic/pocket-casts-ios/pull/404

## Description


This PR
- Generates an image for the latest story and passes it to the share intent.
    
    The image generation uses Android's `view.drawToBitmap()` function. Selected story's sharable composable is converted to `ComposeView` and laid out into `AndroidView` otherwise an exception is thrown: `java.lang.IllegalStateException: View needs to be laid out before calling drawToBitmap()`
 
    It only captures the visible screen.

    The share button click action invokes a callback that generates the bitmap, saves it, and opens it in `share` view. 

- Two other changes includes in this PR:
    - Fix weird state noticed in https://github.com/Automattic/pocket-casts-android/pull/495#issuecomment-1292676258 [94276c7](https://github.com/Automattic/pocket-casts-android/pull/505/commits/94276c73ccee938847ae3533ffe7693c595cade2) - Converts combined flow to state flow on view model scope
    - Introduce `isInteractive` story [052069b](https://github.com/Automattic/pocket-casts-android/pull/505/commits/052069b0e5708c33a2616677304b3b3ce79567f0)

## Testing Instructions

1. Set `END_OF_YEAR_ENABLED` feature flag to `true` in `base.gradle`
2. Run the app
3. When the prompt appears, tap "View My 2022"
4. Tap "Share"
5. Tap "Edit"
6. ✅ Make sure the generated image is correct
8. Go back to the app, and share the story again
9. Select Instagram or Facebook (story)
10. ✅ Check that the image appears correctly in those apps

## Screenshots or Screencast 

Share View| Screenshot
---|----
<img width=320 src="https://user-images.githubusercontent.com/1405144/198588986-1587325e-fc68-45d1-aec0-0011d9972a89.png"/> | <img width=320 src="https://user-images.githubusercontent.com/1405144/198598538-5ca9dee5-bc5f-4265-bc65-d4caf9dbf73c.png"/>

## Checklist
 
#### I have tested any UI changes...
